## Data structures and variables
In this tutorial, we are going to introduce the concept of data structures and variables. A data structure is a way you store data in computer memory, such as a number, a word, or a paragraph. A variable is a value that can change, depending on conditions or on information passed to the program. Using variables to hold data within our code, and show how they can be manipulated to perform whatever task we are intending to do. By the word variables, I mean using names to hold information rather than tapping in numbers as you'd do on a calculator.

### Data structures
There are several built-in core data types in Python, and you can create variable types on your own. Here, we introduce some commonly used built-in variable types.

#### Numbers
Numbers are fairly straightforward. Python’s core object set includes the usual suspects: integers (numbers without a fractional part), floating-point numbers (roughly, numbers with a decimal point in them), and even more exotic types (unlimited-precision “long” integers, complex numbers with imaginary parts, fixed-precision decimals, and sets). Here, we focus on integers and floating-point numbers. Following are some example numbers supported by Python.

In [2]:
# Example of a integer
123

123

In [3]:
# Example of a float
3.1415926

3.1415926

### Logits or boolean
Python Boolean type, bool, has two values only, True and False. This data structure is widely used in branched programs. In Python, it behaves like an integer where 1 represents True and 0 represents False.

In [12]:
True

4

In [11]:
False

False

#### Strings
Strings are used to record textual information as well as arbitrary collections of bytes. They are our first example of what we call a sequence in Python—that is, a positionally ordered collection of other objects. Sequences maintain a left-to-right order among the items they contain: their items are stored and fetched by their relative position. Strictly speaking, strings are sequences of one-character strings; other types of sequences include lists and tuples (covered later). Following are some examples of strings.

In [ ]:
# This is a single character string
'a'

In [ ]:
# This is also a string contains multiple characters
'abc'

#### Lists
The Python list object is the most general sequence provided by the language. It's very flexible. Lists are positionally ordered collections of arbitrarily typed objects, and they have no fixed size. For example, you can make a list using numbers or strings, or mix them together.

In [4]:
# This is a list contains numbers only
[1, 2, 3]

[1, 2, 3]

In [ ]:
# This is a list contains strings only
['a', 'b', 'c']

In [6]:
# This is a list contains numbers and strings
[1, 'a', 2, 'b']

[1, 'a', 2, 'b']

### Variables and assignment of variables.

As mentioned before, a variable is a value that can change, depending on conditions or on information passed to the program. Before you use any of the variables, you need to assign some value to it using '='. Following are some examples of variable asignment.

In [8]:
a = 1
b = [1, 2, 3]
c = a
d = 'abc'

In the code above, the left side of the '=' is the variable name. The variable name can be an arbitrary one, in practice, I recommend you name the variable with its actual meaning, so the code can be easier to understand. The right-hand side is the data stored in the variable name. It can be any data structure you want. Take a = 1 as an example, after this assignment, the number 1 is stored in the variable a, and when you call variable a in future programming, it refers to the assigned value. Let's print this:

In [9]:
print(a)

1


Remember the following few points of the variables:

• Variables are created when they are first assigned values.

• Variables are replaced with their values when used in expressions.

• Variables must be assigned before they can be used in expressions.

• Variables refer to objects and are never declared ahead of time.

Now, let's have a closer look by coding

In [ ]:
a = 1
print(a)
a = 2           # a is reassigned to 2
print(a)


In [ ]:
print(e)        # e is not defined yet, so it will raise an error

Well done, you have knowledge about the variables and data structures now. You can try to assign some values to different variables and print them to see if they follow your expectation.

## Operations



### arithmetic operations
You can use python to do everything that you use a standard calculator for, plus so much more. So let us start with some simple standard operations - adding, subtracting, multiplying and dividing. For these we use +, -, * and / respectively. Have a go at running the snippets of code below to get the hang of it (remember, click on the box so it is highlighted and hit the run button above). Change the numbers and try running again - just remember to start each line flush to the left hand side.

In [ ]:
4+3

In [ ]:
5-2

In [ ]:
5*5

In [ ]:
25/3

Also, these operations can be applied to assigned variables.

In [2]:
a = 5
b = 2

print(a + b)
print(a / b)
c = a * b
print(c)

7
2.5
10


Now you have the hang of this, we can do some more involved maths using multiple operations at once. Brackets () can be used in python in the same way that you're used to with pen and paper. Try the lines below to see how it works, change the numbers and then re-run!

In [ ]:
2*(4+4)

In [ ]:
(5+5)/5

In [ ]:
(5+5)/5 + 2*(4+4)

Now it is your turn. If I have 10 bags, each containing 5 red balls and 3 green balls, can you write an expression that includes brackets () to calculate the total number of balls I have?

Other important operations include powers, and for this we use ** notation. So, for example, the square of the number 4 is written:

In [ ]:
4**2

We can use this to calculate square roots - remembering that the root is a power to the half! So the square root of 2 is:

In [ ]:
2**0.5

Your turn! Can you calculate the lengths of the sides of a square with area 10cm^2?

What might happen if you apply arithmetic operator to some other types of variables? For example, a string? The fact is some operators will be reloaded or reused to complete a different function, and some of them will produce an error. For example, if we add to strings, it will concate them.

In [1]:
a='abc'
b='def'
a+b

'abcdef'

If we multiply a string with another number n, the string will be repeated for n times.

In [2]:
a='abc'
a*3

'abcabcabc'

However, if we multiply two strings, it will produce an error. If your code went run in some cases, try to check if you used a correct data structure.

In [3]:
a='abc'
b='def'
a*b

TypeError: can't multiply sequence by non-int of type 'str'

### Comparison operations

There are five commonly used comparison operators: >, <, ==, >=, <=. These operation will return True if the condition is met and False if the condition is not met. Note that when testing if two numbers are equal, you have to use '==' instead of '=' because '=' is an asignment expression.

In [5]:
4>=3

True

In [6]:
3>4

False

In [7]:
6==6

True

Now, have a try. Using Python to determine which number is bigger, 3.14 or 2.72

To finish off this week, we are going to look at comments and dealing with bugs and errors in our codes.


Firstly, code comments are just messages for the code developers to use to explain the purpose of a line or block of code. They can go anywhere inside your code, even following lines of code you want to compute. The comments are started by the # character and you just place your message after this. When python is executed, if there is a #, it will be ignored along with the rest of the line.

So for example:


In [ ]:
# this is a comment starting at the beginning of the line
# we will calculate the root of a number.
my_number = 10 # set up a variable to hold the number you are going to find square root of.
root = my_number**0.5  # perform the square root.
print('square root of',my_number,'is',root) # printing the answer


You can use a multiline string to make a comment. For example: 

In [ ]:
''' All contents after this mark are comments
All contents here are comments. They won't be executed
print('I will print nothing')

Many programmers write documents using multiline string to make their code easier to understand.
'''

it really is good practice to use comments, particularly for a complicated piece of code. You may write something and not revisit it for months or years, or others may wish to use it later. Comments really help you and others understand what is going on!! SO YOU HAVE BEEN WARNED!!!

Finally debugging code errors - THIS IS REALLY IMPORTANT!!!!!

I guarantee you will spend the vast majority of your time, particularly when learning coding, debugging errors in your code. YOU WILL I PROMISE YOU! Errors and removing them is part of your learning curve. Sometimes they are difficult to find and can drive you crazy! This part is only conquered with experience, so the more you code the better you will get. SO DO ALL THE EXERCISES I GIVE YOU - AND MORE :)

But as a starter, when you write code it is essential to check it produces the correct answer. I very often print to screen many of the executions and check that the code is correct. Very often i do this line by line when possible. So don't write an entire lump of code and then debug. Do it slowly or it will get messy!!!

Anyway, for a practice can you correct the bit of code below where I am trying to find the cube root of a number.

In [ ]:
# this is a comment starting
# we will calculate the cube root of a number.
my_number = 10 # set up a variable to hold the number you are going to find the cube root of.
root = my_number**3  # perform the cube root
print('cube root of ',my_number,' is ',root) # printing the answer

Finally, when coding, you may/will make errors that make no sense to python. In this case python is really good at telling you where you went wrong. Python will stop and tell you the line where it hit a problem and give you information about the problem.

So for example, remember I said don't indent your code: look at what happens when you do this. Indentation is a way of telling a Python interpreter that the group of statements belongs to a particular block of code. It useful in if statements, loops and definations, we will introduce this later and don't indent casually for now. If you run the code python crashes and tells you it's an indent error on the second line!! Very helpful!!! :)


In [ ]:
print('i will show you a typical error')
  print('look when i did not start at the beginning of the line!') # run the code and see what happens.


Run the code above and then fix it. Now practice on the code below, calculating the surface area of a sphere with radius r. Note: Area = 4*pi*r^2 (there are 5 errors!!)


In [ ]:
# some code to calculate the aread of a sphere.
radius = 1.0 # define the radius
 pi = 3.141592 # define pi
area = 4*pi*r**3   calculate the surface area
print('Area =', Area)